packages/fx-core/src/component/generator/templates/templateGenerator.ts

```diff
@@ -0,0 +1,223 @@
+import {
+  CapabilityOptions,
+  MeArchitectureOptions,
+  ProgrammingLanguage,
+  QuestionNames,
+} from "../../../question";
+import { Generator, templateDefaultOnActionError } from "../generator";
+import { convertToLangKey, renderTemplateFileData, renderTemplateFileName } from "../utils";
+import { merge } from "lodash";
+import { GeneratorContext, TemplateActionSeq } from "../generatorAction";
+import { TemplateInfo } from "./templateInfo";
+import { Feature2TemplateName } from "./templateNames";
+
+export interface TemplateGenerator {
+  activate(ctx: Context, inputs: Inputs): boolean;
+  run(
+    ctx: Context,
+    inputs: Inputs,
+    destinationPath: string,
+    actionContext?: ActionContext
+  ): Promise<Result<undefined, FxError>>;
+  getTemplateInfos(
+    ctx: Context,
+    inputs: Inputs,
+    actionContext?: ActionContext
+  ): Promise<Result<TemplateInfo[], FxError>>;
+  post(
+    ctx: Context,
+    inputs: Inputs,
+    destinationPath: string,
+    actionContext?: ActionContext
+  ): Promise<Result<undefined, FxError>>;
+}
+
+export class DefaultTemplateGenerator implements TemplateGenerator {
+  componentName = componentName;
+
+  public activate(ctx: Context, inputs: Inputs): boolean {
+    return Object.keys(Feature2TemplateName).some((feature) =>
+      feature.startsWith(inputs.capabilities)
+    );
+  }
+
+  public getTemplateInfos(
+    ctx: Context,
+    inputs: Inputs,
+    actionContext?: ActionContext
+  ): Promise<Result<TemplateInfo[], FxError>> {
+    const templateName = this.getTemplateName(inputs);
+    const language = inputs[QuestionNames.ProgrammingLanguage] as ProgrammingLanguage;
+    const variables = this.getDefaultReplaceMap(inputs);
+    return Promise.resolve(ok([{ templateName, language, variables }]));
+  }
+
+  @hooks([
+    ActionExecutionMW({
+      enableProgressBar: true,
+      progressTitle: ProgressTitles.create,
+      progressSteps: 1,
+      enableTelemetry: true,
+      telemetryEventName: TelemetryEvent.GenerateTemplate,
+    }),
+  ])
+  public async run(
+    ctx: Context,
+    inputs: Inputs,
+    destinationPath: string,
+    actionContext?: ActionContext
+  ): Promise<Result<undefined, FxError>> {
+    const preResult = await this.getTemplateInfos(ctx, inputs, actionContext);
+    if (preResult.isErr()) return err(preResult.error);
+
+    const templateInfos = preResult.value;
+    for (const templateInfo of templateInfos) {
+      await this.scaffolding(ctx, templateInfo, destinationPath, actionContext);
+    }
+
+    const postRes = await this.post(ctx, inputs, destinationPath, actionContext);
+    if (postRes.isErr()) return postRes;
+
+    return ok(undefined);
+  }
+
+  public post(
+    ctx: Context,
+    inputs: Inputs,
+    destinationPath: string,
+    actionContext?: ActionContext
+  ): Promise<Result<undefined, FxError>> {
+    return Promise.resolve(ok(undefined));
+  }
+
+  public getDefaultReplaceMap(inputs: Inputs): { [key: string]: string } {
```

Comment:
how will developers use/extend this function? Does this need to be public?

Reply:
I make this method a common function, so generators can call it to get a default replace map.